### PR TITLE
Flag improvements/tweaks

### PR DESCRIPTION
**Pull request by imorland**
- Adds the _option_ to allow users to flag their own posts (off by default, maintaining current behaviour)
image

- Increase the size of the flagPostModal from _small_ to _medium_ 
image

- 'off-topic', 'inappropriate' and 'spam' now have an optional extra details textarea for providing additional context, if required.
image

- 'other' - submit remains disabled until extra details are entered
image
image

- Column type change on _flags_ - _reason_details_ changed from _varchar_ to _text_ to allow for longer details input

All changes tested locally, and on giffgaff staging environment.

Requires [flarum/lang-english #148](https://github.com/flarum/lang-english/pull/148)

### CODE REVIEW

File: migrations/2019_10_22_000000_change_reason_text_col_type.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+            $table->text('reason_detail')->change();
+        });
+    },
+
+    'down' => function (Builder $schema) {
+        $schema->table('flags', function (Blueprint $table) {
+            $table->string(['reason_detail'])->change();
+        });
+    }
+];
```
Comment (franzliedke) [inline migrations/2019_10_22_000000_change_reason_text_col_type.php:25]: We decided we would prefer to keep the current `string` field. That should be long enough for a simple explanation; and we can enforce it with a `maxlength` attribute and a validation.

File: src/Api/Serializer/FlagSerializer.php [hunk 2]
```diff
@@ -32,6 +32,7 @@ protected function getDefaultAttributes($flag)
             'reason'       => $flag->reason,
             'reasonDetail' => $flag->reason_detail,
             'createdAt'    => $this->formatDate($flag->created_at),
+            'createdAtDisplay' => date_format($flag->created_at, 'd/m/Y H:i:s'),
```
Comment (franzliedke) [inline src/Api/Serializer/FlagSerializer.php]: Please do this in JS code, as we do in other places. There are helpers like `humanTime` that could be used.

File: src/Listener/AddFlagsApiAttributes.php [hunk 3]
```diff
@@ -44,6 +44,7 @@ public function handle(Serializing $event)
             }
 
             $event->attributes['guidelinesUrl'] = $this->settings->get('flarum-flags.guidelines_url');
+            $event->attributes['canFlagOwn'] = $this->settings->get('flarum-flags.can_flag_own');
```
Comment (franzliedke) [inline src/Listener/AddFlagsApiAttributes.php]: It would be easier to review / merge this quickly, if separate changes were in separate pull requests. :wink: 

@jordanjay29 Do you remember whether we made a decision about this? Is this flag useful enough to include in our bundled extension?

File: js/src/forum/components/FlagPostModal.js [hunk 4]
```diff
@@ -54,12 +60,24 @@ export default class FlagPostModal extends Modal {
...
               </label>
 
               <label className="checkbox">
                 <input type="radio" name="reason" checked={this.reason() === 'spam'} value="spam" onclick={m.withAttr('value', this.reason)}/>
                 <strong>{app.translator.trans('flarum-flags.forum.flag_post.reason_spam_label')}</strong>
                 {app.translator.trans('flarum-flags.forum.flag_post.reason_spam_text')}
+                {this.reason() === 'spam' ? (
+                  <div>
+                    <p><em>{app.translator.trans('flarum-flags.forum.flag_post.optional_details')}</em></p>
+                    <textarea className="FormControl" value={this.reasonDetail()} oninput={m.withAttr('value', this.reasonDetail)}></textarea>
```
Comment (franzliedke) [inline js/src/forum/components/FlagPostModal.js]: When you enter text here, and then choose another category - will the previously entered text appear in that category's text field?

If so, do we want that? :thinking:

File: js/src/forum/components/FlagPostModal.js [hunk 5]
```diff
@@ -77,7 +95,7 @@ export default class FlagPostModal extends Modal {
               className="Button Button--primary Button--block"
               type="submit"
               loading={this.loading}
-              disabled={!this.reason()}>
+              disabled={!this.reason() || this.reason() === 'other' && !this.reasonDetail()}>
```
Comment (franzliedke) [inline js/src/forum/components/FlagPostModal.js]: Let's not add this clause. Disabled buttons without an error message explaining why can confuse users.

**Review by franzliedke — CHANGES_REQUESTED** [review]
Thanks for your initiative!

We discussed this internally, and want to request a bunch of changes.

Most importantly, please avoid changes to the dist JS files, and please don't commit the `composer.lock` file. More comments inline...

File: src/Listener/AddFlagsApiAttributes.php (hunk 3 above)
Comment (jordanjay29) [reply src/Listener/AddFlagsApiAttributes.php]: @franzliedke I don't think we made a decision about it, no. It would be incredibly useful for Flarum.org as well, and I don't think any objections were raised to including it.

File: js/src/forum/components/FlagPostModal.js (hunk 4 above)
Comment (imorland) [reply js/src/forum/components/FlagPostModal.js]: This is intentional. Our thoughts behind it are to save a user from re-typing their explanation if they accidentaly selected the wrong reason.

Open to thoughts and suggestions here...

File: js/src/forum/components/FlagPostModal.js (hunk 5 above)
Comment (imorland) [reply js/src/forum/components/FlagPostModal.js]: In conjunction with the language changes (https://github.com/flarum/lang-english/pull/148/commits/420db03571fd833d95acce512db2ab6e64947ccb#diff-19cc7a6cd74cd8da7215054764db7577L31), this field is marked as 'please specify', whereas the other fields are marked 'optional'

Excuse the giffgaff theme for these screenshots:
image

image

Our reasoning behind this, is that we see in excess of 40% of flags raised as 'other' with no other content. The button is re-enabled once at least 1 character is entered.

File: migrations/2019_10_22_000000_change_reason_text_col_type.php (hunk 1 above)
Comment (imorland) [reply migrations/2019_10_22_000000_change_reason_text_col_type.php:25]: I'd put forward that a field `varchar` with a max of 255 is not enough, as we have found.

Use case: a user flagging a discussion/post as a duplicate of another, and provides a url along with their comments. 

Might I suggest a combination of your suggestion and mine? Let's keep the migration to `text`, but add a `maxlength` variable, exposed to the extension setting? This will allow admin to select their own preference, based on their community needs.

**Comment by imorland** [comment]
Thanks for reviewing @franzliedke, please see some questions inline above :)

I'll get a ticket raised on our side to work on the changes ASAP...

File: migrations/2019_10_22_000000_change_reason_text_col_type.php (hunk 1 above)
Comment (franzliedke) [reply migrations/2019_10_22_000000_change_reason_text_col_type.php:25]: Hmm, I guess that's convincing. cc @luceos @datitisev 

No need for the config for now, I'd say.

File: js/src/forum/components/FlagPostModal.js (hunk 4 above)
Comment (franzliedke) [reply js/src/forum/components/FlagPostModal.js]: Okay, just wanted to ensure it's intentional. If it feels good when using it, we can try this out...

File: js/src/forum/components/FlagPostModal.js (hunk 5 above)
Comment (franzliedke) [reply js/src/forum/components/FlagPostModal.js]: I agree "Other" flags without text content should be prevented, but that should be made clear through an explicit error message.

If people gloss over that small help text, they won't suddenly notice it while wondering why they cannot click the button.

File: src/Listener/AddFlagsApiAttributes.php (hunk 3 above)
Comment (franzliedke) [reply src/Listener/AddFlagsApiAttributes.php]: Okay, let's keep it then.

File: js/src/forum/components/FlagPostModal.js (hunk 5 above)
Comment (franzliedke) [reply js/src/forum/components/FlagPostModal.js]: Some more nuance:
- [This StackOverflow answer](https://ux.stackexchange.com/a/73733/17557) presents some great reasoning for this pattern, with nice examples how this can improve the experience. I can relate to that, but...
- On the other hand, especially in terms of accessibility, [disabled buttons suck](https://axesslab.com/disabled-buttons-suck/). While it is possible to make this work well for e.g. users of screenreaders, it would be considerably more work.

In summary, I would like to request a) an always-enabled submit button, and b) an error message in case "Flag Post" is clicked too early.

We could consider disabling the button visually in addition, to provide the nice experience described in the first link.

**Comment by franzliedke** [comment]
@imorland Have you had a chance to schedule this? Would be cool if we could get this into the upcoming [beta.12 release](https://github.com/flarum/core/milestone/14).

If it helps, you can also split up the various parts of this, so that we can get at least parts of it into the release. The "Allow users to flag their own posts" feature, for example, seems to be done already.

**Comment by imorland** [comment]
@franzliedke Sorry for the delay on this one. I think the latest commit covers everything discussed here...

**Comment by imorland** [comment]
I still need to add `flarum-flags.forum.post.reason-needed` to `lang-english`, I'll do that if you're happy with the rest of the changes :)

File: js/src/forum/components/FlagPostModal.js [hunk 6]
```diff
@@ -46,6 +46,12 @@ export default class FlagPostModal extends Modal {
                 <input type="radio" name="reason" checked={this.reason() === 'off_topic'} value="off_topic" onclick={m.withAttr('value', this.reason)}/>
                 <strong>{app.translator.trans('flarum-flags.forum.flag_post.reason_off_topic_label')}</strong>
                 {app.translator.trans('flarum-flags.forum.flag_post.reason_off_topic_text')}
+                {this.reason() === 'off_topic' ? (
+                  <div>
+                    <p><em>{app.translator.trans('flarum-flags.forum.flag_post.optional_details')}</em></p>
```
Comment (franzliedke) [inline js/src/forum/components/FlagPostModal.js]: These should be marked up as a `<label>`. Styling should be achieved with CSS, not with `<em>` tags. (That said, I am not sold on the italic styling.)

When making this change, we should probably avoid nesting labels, too... :thinking:

File: src/Command/CreateFlagHandler.php [hunk 7]
```diff
@@ -26,18 +28,26 @@ class CreateFlagHandler
      */
     protected $posts;
 
+        /**
```
Comment (franzliedke) [inline src/Command/CreateFlagHandler.php]: There are a few formatting errors in this file's diff. :wink:

**Review by franzliedke — CHANGES_REQUESTED** [review]
Thanks for the quick response! We're getting there... :+1: 

One thing here, and others inline:
- Please don't commit changes to the JS dist files. Our bot will take care of that when merging to master.

**Comment by franzliedke** [comment]
Could you share a screenshot of what the error handling will look like when I try to submit without entering a message?

**Comment by imorland** [comment]
@franzliedke the js dist files were removed in my last commit..

Screenshot of the error:
image

File: src/Command/CreateFlagHandler.php (hunk 7 above)
Comment (imorland) [reply src/Command/CreateFlagHandler.php]: That'll teach me for rushing! Corrected...

**Comment by imorland** [comment]
PR on `lang-english` has now also been updated

File: js/src/forum/components/FlagPostModal.js [hunk 8]
```diff
@@ -46,6 +46,12 @@ export default class FlagPostModal extends Modal {
                 <input type="radio" name="reason" checked={this.reason() === 'off_topic'} value="off_topic" onclick={m.withAttr('value', this.reason)}/>
                 <strong>{app.translator.trans('flarum-flags.forum.flag_post.reason_off_topic_label')}</strong>
                 {app.translator.trans('flarum-flags.forum.flag_post.reason_off_topic_text')}
+                {this.reason() === 'off_topic' ? (
+                  <div>
+                    <label>{app.translator.trans('flarum-flags.forum.flag_post.optional_details')}</label>
```
Comment (franzliedke) [inline js/src/forum/components/FlagPostModal.js]: Now the label needs to be linked to the textarea. :wink:

Comment (franzliedke) [reply js/src/forum/components/FlagPostModal.js]: Also, I'd prefer if we could move this block outside of the surrounding `<label class="checkbox">`.

**Review by franzliedke — CHANGES_REQUESTED** [review]
> the js dist files were removed in my last commit..

Please [don't delete them](https://github.com/flarum/flags/pull/18/files), but rather restore their state before this PR. :wink:

**Comment by franzliedke** [comment]
Thanks for the screenshot! I guess we should strive for moving this kind of form-related error message inline into the form, but that's a more general problem for all of Flarum, so don't worry about it here.

**Comment by franzliedke** [comment]
Actually... the `Modal` base class supports error alerts. Can you try using that for the error message? You can check the `LoginModal` for an example...

File: js/src/forum/components/FlagPostModal.js (hunk 8 above)
Comment (imorland) [reply js/src/forum/components/FlagPostModal.js]: So sorry - extreme tiredness is kicking in, only 5 hours sleep in 2 days :( 

I need to take a break from this for the moment, will pick it back up over the weekend...

Comment (imorland) [reply js/src/forum/components/FlagPostModal.js]: @franzliedke just having a second think, rather than having a label for the textarea, to save some screen real estate, how about using the `textarea`'s `placeholder`

Example:
image

**Comment by imorland** [comment]
Yep, the modal error looks and feels much better for this case. Example:
image

**Comment by franzliedke** [comment]
> So sorry - extreme tiredness is kicking in, only 5 hours sleep in 2 days :(
>
> I need to take a break from this for the moment, will pick it back up over the weekend...

Flarum - stealing your sleep since 2015.

In all seriousness, please don't feel obliged to give us your precious weekend time, or time you could be sleeping. There's still several weeks left until the next release. :smiley:

**Comment by tariqwalji** [comment]
Hi @franzliedke ,

Do you know what else is outstanding in order to get this merged in?

File: js/src/forum/addFlagControl.js [hunk 9]
```diff
@@ -7,7 +7,7 @@ import FlagPostModal from './components/FlagPostModal';
 
 export default function() {
   extend(PostControls, 'userControls', function(items, post) {
-    if (post.isHidden() || post.contentType() !== 'comment' || !post.canFlag() || post.user() === app.session.user) return;
+    if (post.isHidden() || post.contentType() !== 'comment' || !post.canFlag() || post.user() === app.session.user && !Number(app.forum.attribute('canFlagOwn'))) return;
```
Comment (franzliedke) [inline js/src/forum/addFlagControl.js]: I am missing a server-side validation of this flag.

:thinking: 

Also, could we just add that into the server-generated value of each post's `canFlag` attribute? That way, we'd have less logic on the client-side.

**Review by franzliedke — CHANGES_REQUESTED** [review]
Hmm, I must have missed that all the changes were already made. Sorry about that. Just a little thing left over now...

**Comment by imorland** [comment]
@franzliedke moved `canFlag` server side (I don't know why I didn't think of that straight away, sorry)

Also removed `post.user() === app.session.user` from JS, as this check is now carried out BE.

**Review by franzliedke — APPROVED** [review]
Thanks for making the changes! I'll merge.

**Comment by imorland** [comment]
Thank you @franzliedke

**Comment by franzliedke** [comment]
Thank *you*!

I'll make some minor tweaks in master, and also change the translation keys - that felt faster than dumping another round of review on you. :wink:

**Comment by franzliedke** [comment]
@imorland I'm done. I tweaked the translation keys for consistency, removed the custom modal error handling (which basically reimplemented the default, but was less flexible) and fixed two little bugs. :grin: 

Nice work overall!
